language-support/ts/codegen/tests/ts/build-and-lint-test/src/__tests__/test.ts

```diff
@@ -312,3 +312,142 @@ test('create + fetch & exercise', async () => {
   const nonTopLevelContracts = await aliceLedger.query(buildAndLint.Lib.Mod.NonTopLevel);
   expect(nonTopLevelContracts).toEqual([nonTopLevelContract]);
 });
+
+test("multi-{key,query} stream", async () => {
+  const ledger = new Ledger({token: ALICE_TOKEN, httpBaseUrl: httpBaseUrl()});
+
+  function collect<T extends object, K, I extends string, State>(stream: Stream<T, K, I, State>): [State, readonly Event<T, K, I>[]][] {
+    const res = [] as [State, readonly Event<T, K, I>[]][];
+    stream.on('change', (state, events) => res.push([state, events]));
+    return res;
+  }
+  async function create(t: string): Promise<void> {
+    await ledger.create(buildAndLint.Main.Counter, {p: ALICE_PARTY, t, c: "0"});
+  }
+  async function update(t: string, c: number): Promise<void> {
+    await ledger.exerciseByKey(buildAndLint.Main.Counter.Change, {_1: ALICE_PARTY, _2: t}, {n: c.toString()});
+  }
+  async function archive(t: string): Promise<void> {
+    await ledger.archiveByKey(buildAndLint.Main.Counter, {_1: ALICE_PARTY, _2: t});
+  }
+  function sleep(ms: number): Promise<void> {
+        return new Promise(resolve => setTimeout(resolve, ms));
+  }
+  // Add support for comparison queries
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const streamQueriesWithComparison = ledger.streamQueries.bind(ledger) as (t: any, qs: any) => any;
+  const q = streamQueriesWithComparison(
+    buildAndLint.Main.Counter,
+    [{p: ALICE_PARTY, t: "included"},
+     {c: {"%gt": 5}}]);
+  const queryResult = collect(q);
+  const ks = ledger.streamFetchByKeys(
+    buildAndLint.Main.Counter,
+    [{_1: ALICE_PARTY, _2: "included"},
+     {_1: ALICE_PARTY, _2: "byKey"},
+     {_1: ALICE_PARTY, _2: "included"}]);
+  const byKeysResult = collect(ks);
+
+  await create("included");
+  await create("byKey");
+  await create("excluded");
+
+  await update("excluded", 10);
+  await update("byKey", 3);
+  await update("byKey", 6);
+  await update("excluded", 3);
+  await update("included", 2);
+
+  await archive("included");
+  await archive("byKey");
+
+  await create("included");
+
+  await sleep(500);
```

Comment:
Tests are failing about 4/10 times without this. With it I got 40 out of 40 successes. I'm not super happy about it, but this being JS, I don't know of any better way to give the collect callback a chance to run.

Reply:
Can we make a promise out of the callback and `await` that?

Reply:
I suppose we could replace the current `collect` mechanism with a multi-delivery promise, and then explicitly `await` each expected message, rather than handling the whole stream as an array.

Syntactically you'd have a few more method calls and `await` expressions in exchange for this one `sleep`. In terms of runtime behaviour, I believe the test would time out instead of failing, which means that it would be slower for non-timing-related failures and that it will be much more lenient if things get slow.

I'll let you decide if you think the tradeoff is worth it; I have a small preference for the state it's currently in, but it's not a massive change and I'm willing to open a PR for it if you think that's better.

Reply:
Let’s leave it and if it starts flaking we can reconsider.